text/trace/0000-error_flagging.md

```diff
@@ -0,0 +1,86 @@
+# Error Flagging with Status Codes
+This proposal adds two status codes explicitly for use as overrides by the end user, and proposes a canonical mapping of semantic conventions to status codes. This clarifies how error reporting should work in OpenTelemetry.
+ 
+## Motivation
+Error reporting is a fundamental use case for distributed tracing. While we prefer that error flagging occurs within analysis tools, and not within instrumentation plugins, a number of currently supported analysis tools and protocols rely on the existence of an explicit error flag reported from instrumentation. In OpenTelmetry, the error flag is called "status codes."
+ 
+However, there is confusion over the mapping of semantic conventions to status codes, and concern over the subjective nature of errors. Which network failures count as an error? Are 404s an error? The answer is dependent on the situation.
+ 
+There is one major exception. Both application developers and operators have a deep understanding of what constitutes an error in their system. OpenTelemetry must provide a way for these users to control error flagging, and explicitly indicate both when a span should and should not be counted as an error.
+A second exception is supporting analysis tools which require explicit error flagging in the data which they receive. In this case, an operator must be able to apply an error flagging schema at some point during the OTLP data processing pipeline.
+ 
+## Explanation
+The following changes add several missing features required for proper error reporting, and are completely backwards compatible with OpenTelemetry today.
+ 
+### Status Codes
+The following status codes are added to our current  schema.
+ 
+* `DEFAULT` No status has been set. Any errors must be detected by the analysis tool. (This replaces `OK`.)
+* `ERROR` Instrumentation has marked a span as an error. (This replaces `UNKNOWN`.)
+* `OK_OVERRIDE` The user has provided an override. The span should NOT be flagged as an error, regardless of other analysis.
+* `ERROR_OVERRIDE` The user has provided an override. The span SHOULD be flagged as an error, regardless of other analysis.
+ 
+(Note that our current status codes include a long list of error types. We may choose to keep  them, change them, or drop them in favor of a single `ERROR` code. How many error types we have is not relevant to this proposal.)
```

Comment:
I think we should drop the current long list. At the moment I do not see the value of keeping it. The list we have is an attempt to standardize all possible errors into a few buckets of errors that can ever happen but I am not convinced it is a valuable standardization.

I can see some possible ways to use the values from the long list in the backend: to filter or to aggregate. However in my opinion the values are not particularly useful such that filtering or aggregation can give me interesting insights that I can't already get by doing an equivalent filtering or aggregation on the original attribute that resulted in the status code (e.g. the http.status_code). The original attribute likely tells me a better and more precise story of what happened. (And if such original attribute is missing the question is why didn't the developer include it and how am I supposed to understand what the generic error codes mean in the particular situation).

I would like to hear some opposing opinions and anecdotes of people who did use the OpenCensus status code for successful troubleshooting using an analysis tool. I can be convinced in that case.

Reply:
Totally fine! I felt like making a decision on these additional status codes expands the scope of this OTEP unnecessarily. I haven't thought deeply about what the use cases for these codes might be, other than allowing for more nuance in error counting, without the overhead of scanning span attributes and events.